diff --git a/boot/bootm.c b/boot/bootm.c
index 4bdca22ea8c..62436c8d52b 100644
--- a/boot/bootm.c
+++ b/boot/bootm.c
@@ -691,10 +691,15 @@ static int bootm_load_os(struct bootm_headers *images, int boot_progress)
 
 		/* Handle BOOTM_STATE_LOADOS */
 		if (relocated_addr != load) {
+			ulong load_size = load_end - load;
+
 			printf("Moving Image from 0x%lx to 0x%lx, end=0x%lx\n",
 			       load, relocated_addr,
 			       relocated_addr + image_size);
 			memmove((void *)relocated_addr, load_buf, image_size);
+			load = relocated_addr;
+			load_end = relocated_addr + load_size;
+			images->os.load = relocated_addr;
 		}
 
 		images->ep = relocated_addr;
